src/mscorlib/src/System/Globalization/EncodingTable.Unix.cs

```diff
@@ -1,101 +1,162 @@
+            EncodingInfo[] arrayEncodingInfo = new EncodingInfo[encodingDataTableItems.Length];
 
-                case "utf-8":
-                    return CodePageUtf8;
+            for (int i = 0; i < encodingDataTableItems.Length; i++)
+            {
+                CodePageDataItem dataItem = encodingDataTableItems[i];
 
-                case "utf-16":
-                    return CodePageUtf16;
+                arrayEncodingInfo[i] = new EncodingInfo(dataItem.CodePage, dataItem.WebName,
+                    Environment.GetResourceString("Globalization.cp_" + dataItem.CodePage));
+            }
 
-                case "utf-16BE":
-                    return CodePageUtf16BE;
+            return arrayEncodingInfo;
+        }
 
-                case "utf-32":
-                    return CodePageUtf32;
+        internal static int GetCodePageFromName(String name)
+        {
+            if (name == null)
+            {
+                throw new ArgumentNullException("name");
+            }
+            Contract.EndContractBlock();
 
-                default:
-                    return CodePageUtf8;
+            ushort codePage;
+            if (encodingDataTable.TryGetValue(name, out codePage))
+            {
+                return codePage;
             }
+
+            // The encoding name is not valid.
+            throw new ArgumentException(
+                String.Format(
+                    CultureInfo.CurrentCulture,
+                    Environment.GetResourceString("Argument_EncodingNotSupported"), name), "name");
         }
-    
-        [System.Security.SecuritySafeCritical]  // auto-generated
-        unsafe internal static CodePageDataItem GetCodePageDataItem(int codepage) {
-            // TODO: Implement this fully.
-            switch (codepage)
-            {
-                case CodePageUtf7:
-                    return new CodePageDataItem("utf-7", CodePageUtf7, "utf-7", "utf-7", 771);
 
-                case CodePageUtf8:
-                    return new CodePageDataItem("utf-8", CodePageUtf8, "utf-8", "utf-8", 771);
+        internal static CodePageDataItem GetCodePageDataItem(int codepage)
+        {
+            CodePageDataItem dataItem;
 
-                case CodePageUtf16:
-                    return new CodePageDataItem("utf-16", CodePageUtf16, "utf-16", "utf-16", 771);
+            // We synchronize around dictionary gets/sets. There's still a possibility that two threads
+            // will create a CodePageDataItem and the second will clobber the first in the dictionary. 
+            // However, that's acceptable because the contents are correct and we make no guarantees
+            // other than that. 
 
-                case CodePageUtf16BE:
-                    return new CodePageDataItem("utf-16BE", CodePageUtf16BE, "utf-16BE", "utf-16BE", 771);
+            //Look up the item in the hashtable.
+            dataItem = (CodePageDataItem)hashByCodePage[codepage];
```

Comment:
Per my earlier comment, this would be a good reason to consider using the strongly-typed ConcurrentDictionary instead of the object-based Hashtable, to avoid boxing the Int32 every time we look up in the hashtable.  Might even be more efficient overall using a Dictionary with a lock if such lookups are common, to avoid the allocations.